### PR TITLE
feat(core): support argument constraints on permission overrides

### LINKED ISSUE #283: feat(core): Support granular parameter/field-level constraints
### Description
Currently, Corsair checks permissions at the endpoint level (e.g. `messages.post`) and matches exact argument payloads. However, real-world authorization requires field-level validation rules. For example, an agent should be allowed to post messages to `#general` or emails to internal domains (e.g., `*@company.com`) without approval, but require approvals for external DMs or external emails.

### Proposed Architecture
Extend `PluginPermissionsConfig` and the permission evaluator in `packages/corsair/core/permissions/index.ts` to allow rule definitions based on parameter values:
- Add a validator schema or constraint checker (e.g., regex patterns, allowlists/denylists, or custom Zod rules).
- Define constraints inside the permissions configuration:
```json
{
  "plugins": {
    "slack": {
      "mode": "strict",
      "overrides": {
        "messages.post": {
          "policy": "allow",
          "constraints": {
            "channel": { "match": "^#(general|alerts)$" }
          }
        }
      }
    }
  }
}
```
- Integrate this evaluation logic during the `enforcePermission` call before creating/checking pending database records.

### PR DESCRIPTION
**Pull request by thisisanubhav**
## Description

Closes #283. Design was proposed on the issue first ([comment](https://github.com/corsairdev/corsair/issues/283#issuecomment-5368532642)) — the three open questions from it are restated at the bottom, since they're shape decisions rather than implementation details and I'd rather they get reviewed explicitly than merged by default.

Endpoint overrides currently answer *"may the agent call `messages.post` at all?"*. Real authorization needs *"may it call it with **these** arguments?"* — posting to `#general` without approval while an external DM still escalates.

### Config shape

`overrides` values widen from `PermissionPolicy` to a union. **Every existing config keeps compiling and behaves identically** — constraints are opt-in:

```ts
type PermissionOverride =
  | PermissionPolicy                      // unchanged
  | {
      policy: PermissionPolicy;           // applied when all constraints hold
      constraints: Record<string, PermissionConstraint>;
      otherwise?: PermissionPolicy;       // applied when they don't; omit to fall back to the mode matrix
    };

type PermissionConstraint =
  | { match: string }    // regex, string arguments only
  | { equals: unknown }
  | { in: unknown[] }    // allowlist
  | { notIn: unknown[] };
```

The motivating case from the issue:

```ts
slack({
  permissions: {
    mode: 'strict',
    overrides: {
      'messages.post': {
        policy: 'allow',
        constraints: { channel: { match: '^#(general|alerts)$' } },
      },
    },
  },
})
```

Under `strict` a write escalates to `require_approval`. With that constraint, `#general` and `#alerts` run immediately; every other channel still escalates. Constraint keys are dot-notation paths into the endpoint's arguments, so `'message.to': { match: '@corsair\\.dev$' }` reaches nested payloads.

### Placement

Evaluation lives in `evaluatePermission`, which runs *before* any database work in `enforcePermission` — so a constraint-allowed call creates no approval record at all, rather than creating one and immediately resolving it.

### Fails closed

A constraint is **not** satisfied when the argument is absent, is the wrong type for the operator (a number never matches a regex), the regex is malformed, or the constraints map is empty. When constraints don't hold the override contributes `otherwise`, or nothing at all, leaving the mode matrix to decide — so a constraint can only narrow where a loosened policy applies, never widen it.

## Tests

`packages/corsair/tests/permission-constraints.test.ts` — 19 tests:

- each operator, including non-string values against `match` and structural comparison for object operands
- every fail-closed path (missing argument, invalid regex, empty constraint map, unrecognized constraint shape)
- the fallback split: mode matrix by default, `otherwise` when specified
- loosening (`strict` + allowlisted channel → `allow`) and tightening (`cautious` + denylisted recipient → `deny`)
- two end-to-end through `enforcePermission`, asserting the allowed case returns `{ result: 'allow' }` with no approval record

## Checklist

- [x] I have run `pnpm lint` and all checks pass
- [x] I have run `pnpm typecheck` and there are no TypeScript errors
- [x] I have run `pnpm build` and all packages build successfully
- [x] I have run `pnpm test` and all tests pass
- [x] I have added or updated tests where applicable
- [x] I have added or updated necessary documentation

## Screenshots / Demos (if applicable)

No UI or CLI surface — this is a core permission-evaluation change. The working proof is the test run on this PR:

https://github.com/corsairdev/corsair/pull/896/checks

```
Test Suites: 1 skipped, 57 passed, 57 of 58 total
Tests:       1 skipped, 435 passed, 436 total
```

(`postgres-js-database.test.ts` needs a live Postgres and is excluded locally; it is unaffected by this change.) Also clean locally: `tsc --noEmit`, `tsc --build --force`, `biome check` across 113 files, and `validate:docs`.

## Additional Notes

No breaking changes, no new dependencies, no schema or migration changes. Docs updated in `docs/concepts/permissions.mdx` with the operator table, the nested-path and `otherwise` forms, and the fail-closed rules.

**Three decisions I'd like reviewed rather than assumed:**

1. **Fallback on constraint failure** defaults to the mode matrix, not the stated `policy`, so constraints can only narrow. `otherwise` covers the tighten case.
2. **Empty constraints map is unsatisfied.** `{}` reading as "matches everything" would silently apply the loosened policy to every call, which seemed wrong for a security feature — but it is a judgement call.
3. **`match` accepts arbitrary regex from config.** Patterns are developer-authored, but the values they test are agent-controlled, so a catastrophic-backtracking pattern could be made slow by a hostile argument. I documented it with a note to keep patterns anchored. If you'd rather not accept arbitrary regex at all, `equals`/`in`/`notIn` cover the issue's examples and I'm happy to drop `match`.

## Summary by CodeRabbit

* **New Features**
  * Added conditional permission overrides based on endpoint arguments.
  * Supports exact matches, regular expressions, allowlists, denylists, and nested argument paths.
  * Matching constraints apply the configured policy directly, including policies that bypass approval requirements.
  * Added optional fallback policies when constraints do not match.
  * Invalid, missing, or unsupported constraints fail closed.
  * Exported permission constraint and override types for configuration.

* **Documentation**
  * Clarified configuration and fallback behavior for constrained permission overrides.

### CODE REVIEW

**Comment by Mayank-saraswal** [comment]
@greptileai review

**Comment by Mayank-saraswal** [comment]
@greptileai review

**Comment by Mayank-saraswal** [comment]
@greptileai

**Comment by Mayank-saraswal** [comment]
@greptileai review

**Comment by Mayank-saraswal** [comment]
@greptileai

**Comment by Mayank-saraswal** [comment]
@greptileai

**Comment by Mayank-saraswal** [comment]
Addressing the @greptileai
 review — all 9 P1 findings are resolved, the latest of them in 55b52107. Item by item:

1. **Missing values satisfy `notIn`** — fixed. `matchesConstraint` returns `false` for `undefined` before any operator runs, so an omitted argument can never satisfy a constraint (`notIn` included). Regression test added.
2. **Structural equality depends on key order** — fixed. The `JSON.stringify` comparison was replaced with a structural deep-equal that is key-order independent. Tests cover reordered keys under both `equals` and `notIn`.
3. **Malformed constraints crash permission checks** — fixed. `soleOperator` requires exactly one own key that is a recognized operator (misspelled operators, symbol keys, multi-operator shapes, non-array `in`/`notIn` operands all fail closed) — all covered by tests.
4. **Malformed constraint containers still throw** — fixed. `constraintsSatisfied` rejects non-record containers (`null`, primitives) instead of throwing.
5. **Null overrides crash permission checks** — fixed. `resolveOverridePolicy` accepts and fails closed on `null` (the parameter type is widened accordingly).
6. **Exotic objects compare as equal** — fixed. Only plain objects (incl. null-prototype) compare structurally; RegExp/Map/Set/class instances never compare equal by enumerable keys. Dates still compare by time value.
7. **Cyclic graphs compare as equal** — fixed via combined pair keys.
8. **Cycle guard conflates graph nodes** — fixed via right-side partner tracking.
9. **Cycle pairing remains asymmetric** — fixed in 55b52107. The cycle guard now records pairing in **both** directions (`leftPartner` and `rightPartner`), so a one-node self-loop no longer compares equal to a two-node ring regardless of which side the cycle is on. Regression tests cover both directions, plus positive controls that isomorphic cycles still compare equal.

Additional hardening in 55b52107:

- The constraint validators (`matchesConstraint`, `constraintsSatisfied`) now accept `unknown` and validate at runtime — they sit on the untrusted-config boundary where the static annotation cannot be trusted. This removed **every type assertion** from `constraints.ts`; narrowing is done by sound type predicates (`isRecord`, `isPlainObject`, `isUnknownArray`).
- The permission-constraints test suite grew from 44 to 69 tests, adding primitive edge cases (NaN/−0/bigint/symbol semantics), structural edges (explicit `null`/`undefined` members, array-vs-indexed-object, null-prototype objects), cycles through arrays and nested levels, operator operand edges (RegExp-as-pattern rejected, empty lists, NaN in lists), and hostile `resolveArgPath` inputs (empty path segments, prototype members, class instances).

**One limitation flagged for maintainer review rather than silently changed:** DAG-shared cyclic nodes still compare equal to distinct-but-equivalent cyclic counterparts. Enforcing strict aliasing would break value-equality for JSON-parsed configs, since JSON cannot express reference sharing — `{p: shared, q: shared}` vs `{p: {k:1}, q: {k:1}}` are indistinguishable after a JSON round-trip. Happy to switch to fail-closed-on-cyclic-operands instead if that semantics is preferred.

Verification: `tsc --noEmit` (main config, exit 0), `tsup` build, biome, and the full permission test suites pass locally

**Comment by Mayank-saraswal** [comment]
@greptileai review

File: packages/corsair/core/permissions/constraints.ts [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+		// directions: if either this left-hand object or this right-hand
+		// object has been paired with a *different* counterpart in the
+		// current chain, the graphs have non-isomorphic cycle topology.
+		return rightPartner.get(b) === a && leftPartner.get(a) === b;
+	}
+	visited.add(key);
+	const prevRight = rightPartner.get(b);
+	const prevLeft = leftPartner.get(a);
+	rightPartner.set(b, a);
+	leftPartner.set(a, b);
```
[the remark this answers is not in the document]
Comment (Mayank-saraswal) [reply packages/corsair/core/permissions/constraints.ts]: Verified and fixed in e93fc8e4. The counterexample was reproduced first — a two-node ring compared equal to a tail entering a two-node ring in **both** orientations — confirming the pairing guard's overwrite path was the hole.

Rather than patch the pairing guard a fourth time, the semantics changed to what was offered in the previous summary: **cyclic structures now fail closed instead of being compared.**

- `matchesConstraint` rejects the constraint outright when either the resolved value or the operand contains a reference cycle (`containsCycle`, path-based detection — a node counts as cyclic only when it repeats on the *current* path, so values legitimately shared across branches still compare by value).
- This closes the whole class of topology questions at once: this finding, the earlier ring-vs-self-loop cases, and any future non-isomorphic counterexample, because no cyclic comparison happens at all.
- It also closes a related inversion that the pairing guard could never have handled: under `notIn`, an uncomparable cyclic value would have read as "not in the denylist" and satisfied the constraint. Now it fails closed like every other unusable value.
- Rationale: JSON cannot express cycles, so config operands and MCP/HTTP arguments are never cyclic in legitimate use — a cyclic value here is a hand-crafted JavaScript object, and comparing cyclic graphs correctly requires graph isomorphism, not value equality. A security boundary should refuse rather than approximate. The entire cycle-guard machinery (pair keys, partner maps, visited set) is deleted; `deepEqual` is now a plain structural comparator over guaranteed-acyclic inputs.

Tests: the isomorphic-cycle cases now assert fail-closed; new coverage for the `notIn` inversion, cyclic operands on the config side, cyclic entries inside `in`/`notIn` lists, and DAG-shared values still comparing by value. 72/72 in the permission-constraints suite; `tsc --noEmit` exit 0; biome clean. Docs updated (`docs/concepts/permissions.mdx` fail-closed list now states the cyclic rule explicitly).

**Comment by Mayank-saraswal** [comment]
@greptileai

File: packages/corsair/core/permissions/constraints.ts [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+ */
+function containsCycle(root: unknown): boolean {
+	if (!isRecord(root)) return false;
+	const path = new Set<object>();
+	const visit = (v: unknown): boolean => {
+		if (!isRecord(v)) return false;
+		if (path.has(v)) return true;
+		path.add(v);
+		for (const key of Object.keys(v)) {
+			if (visit(v[key])) return true;
```
[the remark this answers is not in the document]
Comment (Mayank-saraswal) [reply packages/corsair/core/permissions/constraints.ts]: Verified and fixed in 2c3f664b. Reproduced first: depth ~10,000 nested objects and arrays threw `RangeError: Maximum call stack size exceeded` inside the recursive `containsCycle` scan — exactly the crash described — while depth 1,000 still worked, confirming the boundary.

Two recursive sites existed, and both are now iterative:

1. **`containsCycle`** — rewritten with an explicit frame stack that preserves the same path enter/leave discipline as the recursive DFS (a node enters the path set when its frame is pushed and leaves when the frame is exhausted). Deep acyclic structures now scan to completion instead of overflowing.
2. **`deepEqual`** — rewritten with an explicit pair stack. Making the comparator iterative (rather than catching `RangeError`) was a deliberate choice: a caught overflow would return `false` from an uncompleted comparison, and under `notIn` that false would invert into a satisfied constraint — "not in the denylist" — which is precisely the fail-open direction this feature exists to prevent. With a non-throwing comparator, every comparison result is complete and truthful, so deep values simply compare by value like any other structure.

Regression tests added: 20,000-deep object and array structures under `equals` and `notIn` (no throw, correct comparison), plus positive controls at normal depth so the iterative rewrite could not have weakened legitimate comparisons. 75/75 in the permission-constraints suite; `tsc --noEmit` exit 0; biome clean; docs updated (`docs/concepts/permissions.mdx` now states that deep structures cannot overflow the check).

**Comment by Mayank-saraswal** [comment]
@greptileai

File: packages/corsair/core/permissions/constraints.ts [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+		if (frame === undefined) break;
+		const [node, children] = frame;
+		const child = children.pop();
+		if (child === undefined) {
+			// Frame exhausted — node leaves the current path.
+			path.delete(node);
+			stack.pop();
+			continue;
+		}
+		if (!isRecord(child)) continue;
```
[the remark this answers is not in the document]
Comment (Mayank-saraswal) [reply packages/corsair/core/permissions/constraints.ts]: Verified and fixed. Reproduced first, with an important refinement to the report: the bypass is **key-order dependent**. The frame stack pops children in reverse insertion order, so `{ a: cyclic, b: undefined }` hid the cycle (the `undefined` child exhausted the frame before the cyclic sibling was visited), while `{ b: undefined, a: cyclic }` was correctly rejected.

The fix matches the suggestion: frame exhaustion is now detected by `children.length === 0` *before* popping, since an own enumerable property whose value is `undefined` is a real child and must not terminate its frame. Both key orders now reject under `notIn` (and under every other operator), as do cyclic operands — no path reaches `deepEqual` unscanned, so there is no hang either.

Regression tests cover both key orders plus an `equals` case, with the DAG-shared positive control still asserting legitimate shared references compare by value. 76/76 in the permission-constraints suite; `tsc --noEmit` exit 0; biome clean.

**Comment by Mayank-saraswal** [comment]
@greptileai

**Comment by Mayank-saraswal** [comment]
@greptileai Fixed the shared-DAG exponential traversal finding in `ec4d5383`.

- `containsCycle` now memoizes fully traversed nodes, so shared acyclic subgraphs are scanned once rather than once per incoming path.
- `deepEqual` now memoizes compared object pairs, avoiding repeated structural comparison of the same shared pair.
- Added a 30-level shared-DAG regression test with observable property reads; it completes with linear work instead of requiring billions of reads.

Validation: 77/77 permission-constraint tests pass, Corsair TypeScript checking passes, Biome passes, and `git diff --check` passes.

**Comment by Mayank-saraswal** [comment]
@greptileai  @ambikeesshh review

**Comment by ambikeesshh** [comment]
@greptileai

**Comment by ambikeesshh** [comment]
@greptileai

**Comment by thisisanubhav** [comment]
Picking this back up after being away — thanks @Mayank-saraswal and @ambikeesshh for carrying it. I verified the two findings from the Aug 23 Greptile pass against the branch as it stands. Both reproduce.

## Array container (`constraints.ts:361`) — fixed in 709631a6

`isRecord` admits arrays, and `constraintsSatisfied` checked only `isRecord`. So `constraints: [{ equals: 'x' }]` reached `Object.entries`, which exposed index `'0'` as an argument path:

```ts
constraintsSatisfied([{ equals: 'x' }], { '0': 'x' })              // was true
evaluatePermission('write', 'strict', { policy: 'allow', constraints: [{ equals: 'x' }] }, { '0': 'x' })
// was 'allow'
```

Rejected explicitly now, with a regression test. 503 tests pass, typecheck and Biome clean.

## Accessor values (`constraints.ts:79`) — real, but not fixable in this file

This one I have not touched, because I do not think it can be fixed in the evaluator, and the fix that would work is broader than this PR should decide unilaterally.

Reproduced on the current branch:

```ts
let n = 0;
const args = { get channel() { return n++ === 0 ? '#general' : '#exec-private'; } };

evaluatePermission('write', 'strict',
  { policy: 'allow', constraints: { channel: { equals: '#general' } } }, args);
// -> 'allow'
args.channel;  // -> '#exec-private'   ← what the endpoint then receives
```

Permission is granted against one value and the provider is called with another. The cause is not in `constraints.ts`: `bind.ts` passes the *same* `args` reference to `enforcePermission` (line 137) and later to the endpoint body, so any re-read can differ. Hardening `resolveArgPath` cannot help — the second read happens in the endpoint, after authorization is complete.

The sound fix is for the binding layer to authorize and execute against the *same materialized values*: take a plain-data snapshot of `args` before `enforcePermission` and pass that snapshot to the endpoint too, so the authorized value and the executed value are identical by construction.

I have deliberately not implemented that here. It changes the argument object for every endpoint call in Corsair, not just constrained ones, and a naive structured clone would break plugins that pass non-cloneable arguments — Buffers for uploads, streams, callbacks. Options as I see them:

1. Snapshot in `bind.ts` for all calls — soundest, widest blast radius.
2. Snapshot only when a constrained override applies to the path — contained, and constraints are the only feature that currently reads argument values during authorization.
3. Treat it as out of scope for #283 and track separately, since the same re-read gap exists for the args recorded on approval records.

I would lean to (2), but it is your call and I would rather ask than reshape the binding layer on my own.

## Standing design question

The three shape questions from [my comment on #283](https://github.com/corsairdev/corsair/issues/283#issuecomment-5368532642) are still open. Worth noting in light of where this PR went: nearly every finding since — key ordering, cycle pairing, exotic objects, shared-DAG traversal, throwing accessors — traces back to deep structural comparison of *object* operands, which exists only because `equals`/`in`/`notIn` accept objects. Restricting them to primitives would remove the deep-equal, the cycle guard, the memoization and this whole class of finding, and still covers every example in #283, which only ever constrains scalars like `channel` and `to`. Happy to make that cut if you want it.
